src/results/aggregator.rs

```diff
@@ -53,41 +71,85 @@ pub async fn aggregate(
+        })
+        .collect();
+
+    let mut outputs = Vec::with_capacity(task_capacity);
 
-    let searx_map_results = searx_map_results.unwrap_or_else(|e| {
-        if debug {
-            log::error!("Error fetching results from Searx: {:?}", e);
+    for task in tasks {
+        if let Ok(result) = task.await {
+            outputs.push(result)
         }
-        HashMap::new()
-    });
+    }
 
-    result_map.extend(ddg_map_results);
+    let mut engine_errors_info: Vec<EngineErrorInfo> = Vec::new();
 
-    searx_map_results.into_iter().for_each(|(key, value)| {
-        result_map
-            .entry(key)
-            .and_modify(|result| {
-                result.add_engines(value.clone().engine());
-            })
-            .or_insert_with(|| -> RawSearchResult {
-                RawSearchResult::new(
-                    value.title.clone(),
-                    value.visiting_url.clone(),
-                    value.description.clone(),
-                    value.engine.clone(),
-                )
-            });
+    let mut initial: bool = true;
+    let mut counter: usize = 0;
+    outputs.iter().for_each(|results| {
+        if initial {
+            match results {
+                Ok(result) => {
+                    result_map.extend(result.clone());
+                    counter += 1;
+                    initial = false
+                }
+                Err(error_type) => {
+                    engine_errors_info.push(EngineErrorInfo::new(
+                        error_type.downcast_ref::<EngineError>().unwrap(),
+                        upstream_search_engines[counter].clone(),
+                    ));
+                    counter += 1
+                }
+            }
+        } else {
+            match results {
+                Ok(result) => {
+                    result.clone().into_iter().for_each(|(key, value)| {
+                        result_map
+                            .entry(key)
+                            .and_modify(|result| {
+                                result.add_engines(value.clone().engine());
+                            })
+                            .or_insert_with(|| -> RawSearchResult {
+                                RawSearchResult::new(
+                                    value.title.clone(),
+                                    value.visiting_url.clone(),
+                                    value.description.clone(),
+                                    value.engine.clone(),
+                                )
+                            });
+                    });
+                    counter += 1
+                }
+                Err(error_type) => {
+                    engine_errors_info.push(EngineErrorInfo::new(
+                        error_type.downcast_ref::<EngineError>().unwrap(),
+                        upstream_search_engines[counter].clone(),
+                    ));
+                    counter += 1
+                }
+            }
+        }
```

Comment:
It would be helpful to provide some explanations about the purpose of this block, `outputs.iter().for_each(...)`. Particularly, it would be good to clarify why we are separating the consideration fi the output is initial.

Reply:
Thanks :heart: for pointing this, I will fix this right away :rocket: :slightly_smiling_face: .